lib/pbench/server/api/resources/generate_key.py

```diff
@@ -0,0 +1,147 @@
+
+
+def generate_key(time_delta: timedelta, user):
+    user_obj = user.get_json()
+    current_utc = datetime.now(timezone.utc)
+    expiration = current_utc + time_delta
+    payload = {
+        "iat": current_utc,
+        "exp": expiration,
+        "user_id": user_obj["id"],
+        "username": user_obj["username"],
+        "audience": Auth.oidc_client.client_id,
+    }
+    try:
+        generated_api_key = jwt.encode(
+            payload, current_app.secret_key, algorithm=Auth._TOKEN_ALG_INT
+        )
+    except (
+        jwt.InvalidIssuer,
+        jwt.InvalidIssuedAtError,
+        jwt.InvalidAlgorithmError,
+        jwt.PyJWTError,
+    ):
+        current_app.logger.exception(
+            "Could not encode the JWT api_key for user: {}", user
+        )
+        abort(
+            HTTPStatus.INTERNAL_SERVER_ERROR,
+            message="INTERNAL ERROR",
+        )
+    return generated_api_key, current_utc, expiration
+
+
+class GenerateKey(ApiBase):
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.POST,
+                OperationCode.CREATE,
+                authorization=ApiAuthorizationType.NONE,
+            ),
+        )
+        self.server_config = config
+
+    def _post(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        Post request for generating a new persistent API key.
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return:
+            Success: 201 with empty payload
+            Failure: <status_Code>,
+                    response_object = {
+                        "message": "failure message"
+                    }
+        """
+        user = Auth.token_auth.current_user()
+        api_key_expire_duration = self.server_config.get(
+            "openid", "api_key_expiration_duration"
+        )
+
+        if user:
+            try:
+                new_key, current_utc, expiration = generate_key(
+                    time_delta=timedelta(minutes=int(api_key_expire_duration)),
+                    user=user,
+                )
+                try:
+                    key = APIKey(
+                        api_key=new_key, created=current_utc, expiration=expiration
+                    )
+                    user.update(api_key=key)
+                    current_app.logger.info(
+                        "New API key registered for user {} and the api_key is {}",
+                        user.username,
+                        new_key,
+                    )
+                except IntegrityError:
+                    current_app.logger.warning("Duplicate api_key got created")
+                    raise APIAbort(
+                        HTTPStatus.CONFLICT,
+                        message="Duplicate api_key created",
+                    )
```

Comment:
> Why would we generate a duplicate key?

Because we got the same request twice at basically the exact same time?  (It _could_ happen....)

> Is this a user error?

Nope; it's an internal "concern" which we need to respond to appropriately, possibly by just returning the existing key value (it's valid, it's "right", and the fact that it's a duplicate doesn't mean it won't work if the user hands it back to us...).

> If you want to restrict users to a single API key (possibly not a bad idea)

As I said elsewhere, I believe that restricting the user to a single key _is_ a bad idea:  if they have some sort of automated system, they'll want the ability to migrate from an old key to a new one as the key values are rotated, which means that, for the migration period, they will need two key values _both_ of which are valid.  (Once the migration is complete, they can revoke the old key and prepare for the next rotation, so they never _need_ more than two keys, but they do need more than one.)

> see `_decode` in various model classes... and it'd really be nice to refactor this up into the model base class so it can be shared, along with `TZDateTime`

Oh, Dog!  Please do not copy that code **_again_** (we have, what, _four_ copies of it, now??).  They _so_ need to be refactored!!